### PR TITLE
Fix Get-WinEvent -FilterHashtable to work with named event data field

### LINKED ISSUE #2327: Get-WinEvent with FilterHashtable generate broken query (filter)
This is UserVoice issue - https://windowsserver.uservoice.com/forums/301869-powershell/suggestions/15969436-get-winevent-filterhashtable-query-is-wrong

Steps to reproduce
------------------
Below test for PrintService log but you can test for any event log.
Replace printer names and server name with actual in your environment.
`# Test 1a`
`# Result: Get-WinEvent : No events were found that match the specified selection criteria.`
`$ef = @{Logname="Microsoft-Windows-PrintService/Operational";Param5="PS-PRN1"}`
`$a=Get-WinEvent -MaxEvents 25 -FilterHashtable $ef -Debug -ComputerName ps1`

`# Test 1b`
`# Query from Test 1a`
`# Result: Get-WinEvent : The specified query is invalid`
`$ed=@"`
`<QueryList>`
`<Query Id="0" Path="microsoft-windows-printservice/operational">`
`<Select Path="microsoft-windows-printservice/operational">`
`    *[([EventData[Data[@Name='Param5']='PS-PRN1']] or [UserData/*/Param5='PS-PRN1'])]`
`</Select>`
`</Query>`
`</QueryList>`
`"@`
`$a=Get-WinEvent -MaxEvents 25 -FilterXml $ed -Debug -ComputerName ps1`

`# Test 2a`
`# Result: Get-WinEvent : No events were found that match the specified selection criteria.`
`$ef = @{Logname="Microsoft-Windows-PrintService/Operational";Param5=@("PS-PRN1","PS-PRN2")}`
`$a=Get-WinEvent -MaxEvents 25 -FilterHashtable $ef -Debug -ComputerName ps1`

`# Test 2b`
`# Query from Test 2a`
`# Result: Get-WinEvent : The specified query is invalid`
`$ed=@"`
`<QueryList>`
`<Query Id="0" Path="microsoft-windows-printservice/operational">`
`<Select Path="microsoft-windows-printservice/operational">`
`    *[([EventData[Data[@Name='Param5']='System.Object[]']] or [UserData/*/Param5='System.Object[]'])]`
`</Select>`
`</Query>`
`</QueryList>`
`"@`
`$a=Get-WinEvent -MaxEvents 25 -FilterXml $ed -Debug -ComputerName ps1`

Expected behavior
-----------------
Get-WinEvent return event or list of events (which really is in the log).

Actual behavior
---------------
1. Get-WinEvent return for 1a and 2a tests (FilterHashtable):
Get-WinEvent : No events were found that match the specified selection criteria.

2.  Get-WinEvent return for 1b and 2b tests (FilterXml):
Result: Get-WinEvent : The specified query is invalid

3. If copy-paste generated filter (query) into Eventvwr.msc then Eventvwr.msc show "The specified query is invalid"

[Help Get-WinEvent ](https://technet.microsoft.com/library/5fe94870-ed6b-4ce2-9500-93846cc65c95%28v=wps.630%29.aspx) say that the parameter's format is `*=<String[]>` but the cmdlet don't process string array in this place absolutely.

Environment data
----------------

```powershell
> $PSVersionTable
PS C:\Windows\system32> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.0.10240.16384
WSManStackVersion              3.0
SerializationVersion           1.1.0.1
CLRVersion                     4.0.30319.42000
BuildVersion                   10.0.10240.16384
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3

PS C:\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.14915.1000
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14915.1000
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

### PR DESCRIPTION
**Pull request by iSazonov**
Now FilterHashtable parameter in Get-WinEvent for named data field:
1. generates a valid query, ex. @{Logname="System";Param="a"}
2. generates a valid query for multiple values, ex.
@{Logname="System";Param="a","b"}

Now I need help in creating tests #2368

Close #2327

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: This is doing two casts, once with the is keyword and again with the explicit cast.  We normally use this pattern to avoid the second cast:
Array dataArray = value as Array;
if (dataArray != null)
{
...

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Using StringBuilder to create the string would be more efficient here.

**Review by PaulHigin — COMMENTED** [review]
We also need tests for this new functionality.

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Again this is for new Issue and RP I suppose.

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Again this is for new Issue and RP I suppose.

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Paul thanks for code review!
I agree but the entire file uses the old technique!!! I suppose that it makes no sense to change the code in a single place, and there should be a separate issue and PR.

Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Good point.  I didn't realize this pattern was used in other places in the code, but I now see it is used in five other places.  This must be old code because it is not a good or common pattern that we use.  I feel it is worth changing in these five places.  Can you make this part of your change?  At the very least I feel that your new change should use the common pattern even if we create a new issue to clean up the rest of the code.

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: > "((EventData[Data[@Name='{0}']='{1}']) or (UserData/*/{0}='{1}'))"

Let's not repeat this twice. Maybe it's better to hold this format in a const variable.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: > At the very least I feel that your new change should use the common pattern even if we create a new issue to clean up the rest of the code.

I agree with @PaulHigin. Using the recommended pattern in your new code can serve as a good example for anyone who'd come later to fix the rest of code.

**Comment by daxian-dbw** [comment]
@iSazonov I have manually verified your code changes and it works as expected, thanks for fixing it! Now we are waiting on you to address the PR comments.
At the same time, @JamesWTruher is following up on the `.evtx` test file (thanks Jim!). Once we have that test file ready and merged, you can add test for this PR.

**Comment by JamesWTruher** [comment]
I was able to construct a set of tests for this - it required an EVTX file - I'll create a PR for that (the test which uses filterhash is commented out - take a look, and if it's good, you can uncomment the test. I don't want to mark it as pending because on linux it means that tests would be marked both pending and skipped which is not allowed.

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Fixed!

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Fixed!

**Comment by iSazonov** [comment]
@JamesWTruher Great test! Thanks!

Required addition.
This test checks only the first part of my fixes.
>$filter = @{ path = "$eventLogFile"; Param2 = "Windows x64"}
The second part would look like this
>$filter = @{ path = "$eventLogFile"; Param2 = **"Windows x64","Windows x32"**}
as the named parameter by definition are String[].

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Thanks!

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Would you like to address this comment as well (at least in your newly added method `HandleNamedDataHashValue`)?

**Comment by daxian-dbw** [comment]
Thanks @iSazonov for the updates. The EVTX file is now available, please add another test to cover all your changes. Now we are waiting on you to address the `StringBuilder` comment and add the additional test.

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: Thanks!

File: src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs]: @daxian-dbw I oppose for two reasons:
1. As mentioned, the code is very old and needed to be revised entirely. I don't like the idea of disguising one hole out of thirty.
2. Soon I plan to expand the cmdlet and add support Suppress. To do this, I need to completely rewrite the code that builds a query. #2402  
So I don't want to do the work twice. To solve #2402  I will already use the new technique.

**Comment by JamesWTruher** [comment]
Please feel free to add these tests in your PR, since the test code is now all merged, you should be able to add tests you think are good

---
In reply to: [250667902](https://github.com/PowerShell/PowerShell/pull/2373#issuecomment-250667902) [](ancestors = 250667902)

**Comment by daxian-dbw** [comment]
@iSazonov thanks for your replies to the review comments. As for the additional test, please see @JamesWTruher 's comment.

**Comment by iSazonov** [comment]
@JamesWTruher I can add second test, but it requires additional events in the evtx file. Now there are only three events, only one is suitable for new test and need at least two.
>$filter = @{ path = "$eventLogFile"; Param2 = "Windows x64","Windows x32"}

- here for the "Windows x64" there are a event and for "Windows x32" a event does not exists in evtx.

**Comment by JamesWTruher** [comment]
I think you can use DriverName (it's in UserData.SetupInstallPrinterDriver) - it's not Param, but it should still work:
```
<Event xmlns='http://schemas.microsoft.com/win/2004/08/events/event'>
...
<UserData>
<SetupInstallPrinterDriver xmlns='http://manifests.microsoft.com/win/2005/08/windows/printing/spooler/core/events'>
...
<DriverName>Microsoft enhanced Point and Print compatibility driver</DriverName>
...
</SetupInstallPrinterDriver>
</UserData>
</Event>

<Event xmlns='http://schemas.microsoft.com/win/2004/08/events/event'>
...
<UserData>
<SetupInstallPrinterDriver ...
<DriverName>Remote Desktop Easy Print</DriverName>
... 
</SetupInstallPrinterDriver>
</UserData>
</Event>
```

**Comment by daxian-dbw** [comment]
@iSazonov Please see @JamesWTruher's comment about using 'DriverName' to test passing a string array value like `$filter = @{ path = "$eventLogFile"; DriverName = "Remote Desktop Easy Print", "Microsoft enhanced Point and Print compatibility driver" }`.
And also, could you please rebase your branch? Your PR contains some merge commits.

**Comment by iSazonov** [comment]
1. Sorry for the delay the #2425 it stopped me. Now all test ready.
2. Sorry for the refuse I only learn git. Now I made rebase.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Get-WinEvent.Tests.ps1
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Get-WinEvent.Tests.ps1]: > @{ path = "$eventLogFile"; ; PackageAware= ...

A small typo at "; ;". Can you please fix it?

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

**Comment by daxian-dbw** [comment]
@iSazonov Thanks for the update! It look mostly good except for one small typo (please see my comment above). Once you fix the typo, the PR will be ready to merge.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Get-WinEvent.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Get-WinEvent.Tests.ps1]: Fixed!

**Comment by iSazonov** [comment]
@daxian-dbw Small question about planned support of <Suppress> and use StringBuilder. In fact, we need to build a **XML** query: is it acceptable to use System.Xml.XmlWriter?

**Review by daxian-dbw — APPROVED** [review]
LGTM.

**Comment by daxian-dbw** [comment]
@iSazonov Performance difference between `XmlWriter` and `StringBuilder` shouldn't be a consideration here, and they both will avoid generating a bunch of new strings like the current code does (Yay! :)). So you should pick the solution that you think is more maintainable.

`XmlWriter` make it easier to not generate any malformed XML queries, especially when the query is complex, but on the other hand it possibly will make the code less concise, for example, you may end up repeating following code snippet in many places. So if you decide to go with `XmlWriter`, please try avoid repeat as much as possible 😄 (for example, maybe put the code snippet below in a method that you can call whenever you need to generate a `<Query>` element)
```powershell
$xmlWriter.WriteStartElement("QueryList")
$xmlWriter.WriteStartElement("Query")
$XmlWriter.WriteAttributeString("Id", $Id.ToString())
$XmlWriter.WriteAttributeString("Path", $Path)

$xmlWriter.WriteStartElement("Select")
$XmlWriter.WriteAttributeString("Path", $Path)
```

**Comment by daxian-dbw** [comment]
@iSazonov Thanks again for fixing this issue with Get-WinEvent!

**Comment by iSazonov** [comment]
@daxian-dbw Thanks for help and advices! This is my first code after 20 years 😊

**Pull request merged**
